Cleanup

Hurricos opened the pull request:
Branch may still need some testing, but the changes are minor enough and are mostly installation-related.

Some various cleanup:

- Repair `./install.sh [remove]`, touching up various logic errors and other bugs, and adding a new option `uninstall` which avoids removing packages as well.

- Move the PiFmRds repository to `/usr/local/src/`. As fishy of a practice it might seem to clone a git repository there, it appears to work just fine. (Also, move the `pi_fm_rds` binary to `/usr/local/bin/`.)

XXX: We might want to peg a commit to clone from PiFmRds, in case of future changes. Otherwise, changes in the PiFmRds repository may break this repo.

Hurricos committed:
Make uninstall work in some more cases

Hurricos committed:
Fix removing ipv6 from blacklist.conf

Also, from now on, label our blacklists, and only remove labeled
blacklists - to make sure we don't destroy a blacklist a user created.

Hurricos committed:
Add option to uninstall without removing packages

A user can now run `sudo ./install.sh uninstall` instead, which has
the effect of removing the installation, but without removing packages
that `mpradio` happens to depend upon. This is good if you want to keep
things like `bluez`, `git` etc.

XXX: Maybe instead, rebuild dependency management around `dpkg`?

Hurricos committed:
Fix some unnecessary chmodding, and a handle() bug involving chmod

chmodding is no longer necessary: github is smart enough these days to
track executible permissions on files, and we are using these
executibles on a very special set of systems (Raspberry Pi's) which
should not have any issues with those permissions when cloning.

Hurricos committed:
Store PiFmRds in /usr/local/src to ease dependance on /home/pi

Hurricos committed:
Make install / uninstall message more generic

(This is a minor commit.)

Hurricos committed:
Put pi_fm_rds binary into /usr/local/bin/

This is where it belongs, according to the File Hiearchy Standard.

I think.

morrolinux commented:
Great job man! This really needed to be done since a while.
Just a couple of observations:
1) I don't remember exactly why snd_bcm2835 needed to be blacklisted. I think it caused troubles with pulseaudio (or maybe it still does, even by itself?) because I see you haven't blacklisted it in the new installer. Could you argument a bit the choice? 
2) Please note that need2recompile.sh would need to copy the new compiled binary to /usr/local/bin/pi_fm_rds for this to work, or the old executable will still be called, thus no effect would be seen. (It used to work in the past since we were calling it from the sources folder, but I think you already got it :) 

* just out of curiosity, do (how) you test each aspect of code you push in PR?

Hurricos commented:
1) I have not actually removed anything: the effective line to _add_ the blacklist was already commented out. The only thing that wasn't was the removal using `sed`, but no need to remove if it's never added. Also, having the control being based on the presence of the `snd_bcm2835` blacklist was causing issues since it was never being added, so never being grepped. I found 33 copies of `blacklist ipv6` in my `blacklist.conf`... :laughing:

2) Thank you *very* much for the review. I would _never_ have caught that, I think. Pushing the change.

3) I have been somewhat inconsistent about testing. In this case, I had wanted to install, uninstall, reinstall, then plug in a drive to test out RDS with my father's truck, and then connect to bluetooth to listen in via a portable radio receiver I have. However, last night, I had no access to the truck or physical access to the Pi; I did not want to wake anyone up by trying to get to the office (it is currently being used simultaneously as a CUPS server), so all I did was repeatedly install and uninstall to see what happened (at which point I noticed the holes in the uninstaller). Since I caught and replaced all instances of /home/pi paths being used - there were very few, since the executables tend to be called from only one place each - I figured I was finished.

Hurricos committed:
Make sure to copy pi_fm_rds upon recompile

That is, when need2recompile.sh is triggered, make sure to move the
recompiled binary.

morrolinux commented:
1) lol I haven't thought about that
2) np that's why I check the code before merging :)
3) alright that make sense!!

merging now!
